lib/vnet/dns/dns.go

```diff
@@ -0,0 +1,402 @@
+	// error, the question will be forwarded upstream.
+	ResolveAAAA(ctx context.Context, domain string) (Result, error)
+}
+
+// Result holds the result of DNS resolution.
+type Result struct {
+	// A is an A record.
+	A [4]byte
+	// AAAA is an AAAA record.
+	AAAA [16]byte
+	// NXDomain indicates that the requested domain is invalid or unassigned, this is an authoritative answer.
+	NXDomain bool
+	// NoRecord indicates the domain exists but the requested record type doesn't, this is an authoritative
+	// answer.
+	NoRecord bool
+}
+
+// UpstreamNameserverSource provides the current set of upstream nameservers.
+type UpstreamNameserverSource interface {
+	// UpstreamNameservers should return the current set of upstream nameservers, requests that cannot be
+	// resolved will be forwarded to these addresses.
+	UpstreamNameservers(context.Context) ([]string, error)
+}
+
+// Server is a DNS server.
+type Server struct {
+	resolver                 Resolver
+	upstreamNameserverSource UpstreamNameserverSource
+	messageBuffers           sync.Pool
+	slog                     *slog.Logger
+}
+
+// NewServer returns a DNS server that handles the details of the DNS protocol and asks [resolver] to answer
+// DNS questions. If [resolver] has no answer, requests will be forwarded to the upstream nameservers provided
+// by [upstreamNameserverSource].
+func NewServer(resolver Resolver, upstreamNameserverSource UpstreamNameserverSource) (*Server, error) {
+	return &Server{
+		resolver:                 resolver,
+		upstreamNameserverSource: upstreamNameserverSource,
+		messageBuffers: sync.Pool{
+			New: func() any {
+				buf := make([]byte, maxUDPDNSMessageSize)
+				return &buf
+			},
+		},
+		slog: slog.With(teleport.ComponentKey, "VNet.DNS"),
+	}, nil
+}
+
+// getMessageBuffer returns a buffer of size [maxUDPMessageSize]. Call [returnBuf] to return the buffer to the
+// shared pool. Use this to avoid large allocations on all DNS messages.
+func (s *Server) getMessageBuffer() (buf []byte, returnBuf func()) {
+	buf = *s.messageBuffers.Get().(*[]byte)
+	buf = buf[:cap(buf)]
+	return buf, func() {
+		s.messageBuffers.Put(&buf)
+	}
+}
+
+// HandleUDP reads and handles a single UDP message from [conn] and writes the response back to [conn].
+// This will be called by VNet code.
+func (s *Server) HandleUDP(ctx context.Context, conn net.Conn) error {
+	buf, returnBuf := s.getMessageBuffer()
+	defer returnBuf()
+
+	n, err := conn.Read(buf)
+	if err != nil {
+		return trace.Wrap(err, "failed to read from UDP conn")
+	}
+	if n >= maxUDPDNSMessageSize {
+		return trace.BadParameter("Dropping UDP message that is too large")
+	}
+	buf = buf[:n]
+
+	return trace.Wrap(s.handleDNSMessage(ctx, conn.RemoteAddr().String(), buf, conn))
+}
+
+// ListendAndServeUDP reads all incoming UDP messages from [conn], handles DNS questions, and writes the
+// responses back to [conn].
+// This is not called by VNet code and basically exists so we can test the resolver outside of VNet.
```

Comment:
Smells a bit fishy, like there should be a better way to simulate calling `HandleUDP` itself, but I don't know it. Maybe more will be revealed in the next PR.